internal/rbd/controllerserver.go

```diff
@@ -262,27 +258,121 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+	rbdSnap := &rbdSnapshot{}
+	if acquired := cs.SnapshotLocks.TryAcquire(snapshotID); !acquired {
+		klog.Infof(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), snapshotID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, snapshotID)
+	}
+	defer cs.SnapshotLocks.Release(snapshotID)
+
+	err := genSnapFromSnapID(ctx, rbdSnap, snapshotID, cr)
+	if err != nil {
+		if _, ok := err.(util.ErrPoolNotFound); ok {
+			klog.Errorf(util.Log(ctx, "failed to get backend snapshot for %s: %v"), snapshotID, err)
+			return status.Error(codes.InvalidArgument, err.Error())
+		}
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	// update parent name(rbd image name in snapshot)
+	rbdSnap.RbdImageName = rbdSnap.RbdSnapName
+	// create clone image and delete snapshot
+	err = cloneRbdImageFromSnapshot(ctx, rbdVol, rbdSnap)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone rbd image %s from snapshot %s: %v"), rbdSnap, err)
+		return err
+	}
+
+	err = rbdVol.flattenRbdImage(ctx, cr)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Infof(util.Log(ctx, "create volume %s from snapshot %s"), rbdVol.RequestName, rbdSnap.RbdSnapName)
+	return nil
+}
+
+func (cs *ControllerServer) createBackingImage(ctx context.Context, cr *util.Credentials, rbdVol *rbdVolume, rbdSnap *rbdSnapshot) error {
 	var err error
 
+	var j = &journal.Connection{}
+	j, err = volJournal.Connect(rbdVol.Monitors, cr)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer j.Destroy()
+
 	if rbdSnap != nil {
-		err = restoreSnapshot(ctx, rbdVol, rbdSnap, cr)
+		err = cs.createVolumeFromSnapshot(ctx, cr, rbdVol, rbdSnap.SnapID)
 		if err != nil {
 			return err
 		}
-
 		klog.V(4).Infof(util.Log(ctx, "created volume %s from snapshot %s"), rbdVol.RequestName, rbdSnap.RbdSnapName)
-		return nil
+	} else {
+		err = createImage(ctx, rbdVol, cr)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to create volume: %v"), err)
+			return status.Error(codes.Internal, err.Error())
+		}
 	}
 
-	err = createImage(ctx, rbdVol, cr)
+	klog.V(4).Infof(util.Log(ctx, "created volume %s backed by image %s"), rbdVol.RequestName, rbdVol.RbdImageName)
+
+	defer func() {
+		if err != nil {
+			if _, ok := err.(ErrFlattenInProgress); !ok {
+				if deleteErr := deleteImage(ctx, rbdVol, cr); deleteErr != nil {
+					klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s with error: %v"), rbdVol, deleteErr)
+				}
+			}
+		}
+	}()
+	err = rbdVol.getImageID()
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed to create volume: %v"), err)
+		klog.Errorf(util.Log(ctx, "failed to get volume id %s: %v"), rbdVol, err)
 		return status.Error(codes.Internal, err.Error())
 	}
 
-	klog.V(4).Infof(util.Log(ctx, "created volume %s backed by image %s"), rbdVol.RequestName, rbdVol.RbdImageName)
+	err = j.ReserveImageID(ctx, rbdVol.JournalPool, rbdVol.ReservedID, rbdVol.ImageID, cr)
```

Comment:
We may need to move this further up, like for snapshots, and ensure when we find an existing image, this data is already stored.